selfservice/flow/profile/handler.go

```diff
@@ -100,24 +100,38 @@ func (h *Handler) initUpdateProfile(w http.ResponseWriter, r *http.Request, ps h
 		return
 	}
 
+	traitsSchema, err := h.c.IdentityTraitsSchemas().FindSchemaByID(s.Identity.TraitsSchemaID)
+	if err != nil {
+		h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
+		return
+	}
+
 	a := NewRequest(h.c.SelfServiceProfileRequestLifespan(), r, s)
-	a.Form = form.NewHTMLFormFromJSON(urlx.CopyWithQuery(
+	a.Form, err = form.NewHTMLFormFromJSONSchema(urlx.CopyWithQuery(
 		urlx.AppendPaths(h.c.SelfPublicURL(), PublicProfileManagementUpdatePath),
 		url.Values{
 			"request": {a.ID.String()},
 		},
-	).String(), json.RawMessage(s.Identity.Traits), "traits")
-	a.Form.SetCSRF(h.csrf(r))
-
-	traitsSchema, err := h.c.IdentityTraitsSchemas().FindSchemaByID(s.Identity.TraitsSchemaID)
+	).String(), traitsSchema.URL, "traits")
 	if err != nil {
 		h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
 		return
 	}
+
+	a.Form.SetValuesFromJSON(json.RawMessage(s.Identity.Traits), "traits")
+	a.Form.SetCSRF(h.csrf(r))
+
 	if err := a.Form.SortFields(traitsSchema.URL, "traits"); err != nil {
 		h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
 		return
 	}
+
+	for i := range a.Form.Fields {
+		if a.Form.Fields[i].IsIdentifier {
```

Comment:
If the compiler was created in here (above NewHTMLFormFromJSON ) there would be no need to add another field to `form.Fields`. Instead you would be simply setting `Disabled: true` [here](https://github.com/ory/kratos/pull/238/files#diff-081f047cbec5f8cc6a3c7350a79d9815R130-R136) and could remove this completely.

This is because now you explicitly define the compiler for `NewHTMLFormFromJSON`, thus deciding if `schema.IsIdentifierFlag` should be set in the result or not.